src/Hedgehog/Plutus/BadGen.hs

```diff
@@ -0,0 +1,165 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Hedgehog.Plutus.BadGen where
+
+import Data.Bifunctor (Bifunctor (bimap))
+
+import GHC.Float (int2Double)
+import GHC.Generics qualified as GHC
+
+import Generics.SOP (
+  All,
+  All2,
+  HApInjs (hapInjs),
+  HCollapse (hcollapse),
+  HPure (hcpure),
+  HSequence (hsequence'),
+  I,
+  K (K),
+  NP,
+  Proxy (Proxy),
+  SOP (SOP),
+  hcmap,
+  hctraverse,
+  type (:.:) (Comp),
+ )
+import Generics.SOP.GGP (GCode, GTo, gto)
+
+import Hedgehog qualified
+import Hedgehog.Gen qualified as Hedgehog
+
+import Generics.SOP.Dict (Dict (Dict), mapAll)
+import Hedgehog.Plutus.Generics (Generically, Simple (Simple))
+import Hedgehog.Plutus.TestData (
+  Bad,
+  EitherOr (EitherOr),
+  ShouldBeNatural (MightBeNegative),
+  ShouldEqual (MightNotEqual),
+  Shouldn'tExist (MaybeExists),
+  TestData (Good, generalise),
+  test,
+ )
+
+class TestGen a where
+  mutationPoints :: Proxy a -> Double
+
+badGen :: (Hedgehog.MonadGen m, TestData a) => m a -> m (Bad a)
+badGen = Hedgehog.mapMaybeT (either Just (const Nothing) . test)
+
+instance TestGen (Simple a) where
+  mutationPoints _ = 0
+
+simpleGen :: (Functor m) => m a -> m (Simple a)
+simpleGen = fmap Simple
+
+instance TestGen (ShouldEqual val a) where
+  mutationPoints _ = 1
+
+shouldEqualGen :: (Functor m) => m a -> m (ShouldEqual val a)
+shouldEqualGen = fmap MightNotEqual
+
+instance (TestGen good) => TestGen (EitherOr bad good) where
+  mutationPoints _ = 1 + mutationPoints (Proxy @good)
+
+eitherOrGen ::
+  forall m bad good.
+  (TestGen good, Hedgehog.MonadGen m) =>
+  m bad ->
+  m good ->
+  m (EitherOr bad good)
+eitherOrGen bad good =
+  EitherOr
+    <$> Hedgehog.frequency
+      [ (ceiling (1 / (points + 1)), Left <$> bad)
+      , (ceiling (points / (points + 1)), Right <$> good)
+      ]
+  where
+    points = mutationPoints (Proxy @good)
+
+instance TestGen (Shouldn'tExist a) where
+  mutationPoints _ = 1
+
+shouldntExistGen :: (Functor m) => m a -> m (Shouldn'tExist a)
+shouldntExistGen = fmap (MaybeExists . Just)
+
+instance TestGen (ShouldBeNatural a) where
+  mutationPoints _ = 1
+
+shouldBeNaturalGen :: (Functor m) => m i -> m (ShouldBeNatural i)
+shouldBeNaturalGen = fmap MightBeNegative
+
+instance (All2 TestGen (GCode a)) => TestGen (Generically a) where
+  mutationPoints _ = nsPoints (Proxy @(GCode a))
+    where
+      nsPoints :: forall xss. (All2 TestGen xss) => Proxy xss -> Double
+      nsPoints _ =
+        mean
+          . hcollapse
+          . id @(->) @(NP (K Double) xss)
```

Comment:
I found a way to avoid the type hinting id thing
```haskell
npPoints :: forall k (xs :: [k]) . (All TestGen xs) => Proxy xs -> Double
npPoints _ =
  sum
    . hcollapse @k @[k] @NP @xs @Double
    $ hcpure (Proxy @TestGen) genPoints
  where
    genPoints :: forall x. (TestGen x) => K Double x
    genPoints = K $ mutationPoints (Proxy @x)

```

Reply:
Oh I commented this on the wrong npPoints, sorry about that

Reply:
```haskell
      nsPoints :: forall k (xss :: [[k]]). (All2 TestGen xss) => Proxy xss -> Double
      nsPoints _ =
        mean
          . hcollapse @k @[k] @NP @xss @Double
          $ hcpure (Proxy @(All TestGen)) npPoints'
```
works for nsPoints though

Reply:
I had it with `@k @[k]` and got a ghc panic
```
src/Hedgehog/Plutus/BadGen.hs:97:36: error:ghc: panic! (the 'impossible' happened)
  (GHC version 9.2.5:
    No skolem info:
  [k_afC9[sk:3]]
  Call stack:
      CallStack (from HasCallStack):
        callStackDoc, called at compiler/GHC/Utils/Panic.hs:181:37 in ghc:GHC.Utils.Panic
        pprPanic, called at compiler/GHC/Tc/Errors.hs:2912:17 in ghc:GHC.Tc.Errors

Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug
```